assembly/src/parsers/ast/serialization.rs

```diff
@@ -0,0 +1,1165 @@
+            Self::RPHash => target.write_opcode(OpCode::RPHash),
+            Self::RPPerm => target.write_opcode(OpCode::RPPerm),
+            Self::MTreeGet => target.write_opcode(OpCode::MTreeGet),
+            Self::MTreeSet => target.write_opcode(OpCode::MTreeSet),
+            Self::MTreeCWM => target.write_opcode(OpCode::MTreeCWM),
+
+            // ----- exec / call ----------------------------------------------------------------------
+            Self::ExecLocal(v) => {
+                target.write_opcode(OpCode::ExecLocal);
+                target.write_u32(*v);
+            }
+            Self::ExecImported(imported) => {
+                target.write_opcode(OpCode::ExecImported);
+                target.write_string(imported);
+            }
+            Self::CallLocal(v) => {
+                target.write_opcode(OpCode::CallLocal);
+                target.write_u32(*v);
+            }
+            Self::CallImported(imported) => {
+                target.write_opcode(OpCode::CallImported);
+                target.write_string(imported);
+            }
+        }
+    }
+}
+
+pub struct ByteReader;
+
+impl ByteReader {
+    pub fn get_bool(bytes: &mut &[u8]) -> Result<bool, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(1);
+        *bytes = remaining_bytes;
+        let result = u8::from_le_bytes(result.try_into().unwrap());
+        u8_to_bool(result)
+    }
+
+    pub fn get_u8(bytes: &mut &[u8]) -> Result<u8, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(1);
+        *bytes = remaining_bytes;
+        let result = u8::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
+
+    pub fn read_u8(bytes: &mut &[u8]) -> Result<u8, SerializationError> {
+        Ok(bytes[0])
+    }
+
+    pub fn get_u16(bytes: &mut &[u8]) -> Result<u16, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(2);
+        *bytes = remaining_bytes;
+        let result = u16::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
+
+    pub fn get_u32(bytes: &mut &[u8]) -> Result<u32, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(4);
+        *bytes = remaining_bytes;
+        let result = u32::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
+
+    pub fn get_u64(bytes: &mut &[u8]) -> Result<u64, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(8);
+        *bytes = remaining_bytes;
+        let result = u64::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
+
+    pub fn get_usize(bytes: &mut &[u8]) -> Result<usize, SerializationError> {
+        let (result, remaining_bytes) = bytes.split_at(8);
+        *bytes = remaining_bytes;
+        let result = usize::from_le_bytes(result.try_into().unwrap());
+
+        Ok(result)
+    }
```

Comment:
We shouldn't need this one as we shouldn't be writing `usize` into the serialized AST in the first place. But, we probably should add something like `read_opcode` which would read and return a valid opcode (we should be able to do this in a few lines of code).